src/Pact/Types/Principal.hs

```diff
@@ -156,22 +164,47 @@ principalParser (getInfo -> i) = kParser
       eof'
       pure $ U n h
 
--- | Given a pact guard, convert to a principal type
---
-guardToPrincipal :: Guard (Term Name) -> Eval e Principal
-guardToPrincipal = \case
-  GPact (PactGuard pid n) -> pure $ P pid n
+    cParser = do
+      char' 'c'
+      char' ':'
+      h <- base64UrlHashParser
+      eof'
+      pure $ C h
+
+
+-- | Given a pact guard, convert to a principal type.
+-- Parameterized for use with non-Eval contexts.
+guardToPrincipal
+    :: Monad m
+    => (forall f . Traversable f => f n -> m (f PactValue))
+    -- ^ pact value coerce
+    -> (Int -> m ())
+    -- ^ gas charger
+    -> Guard n
+    -> m Principal
+guardToPrincipal toPactValue chargeGas = \case
```

Comment:
I don't actually have much of a problem with this signature as is - it's really clever, but it does look like we need to tease out the subset of guards that are meaningful Pact Values so we don't have to enforce anything.

Reply:
> it does look like we need to tease out the subset of guards that are meaningful Pact Values so we don't have to enforce anything

What does "meaningful" mean here? Term -> PactValue is unsafe so this has to be monadic. In a non-Eval case however `n` is `PactValue` so the function is simply `pure`.

Reply:
We had a discussion on slack about it. The previous PR for principals had this function unused and exposed for use in Chainweaver/haskell clients, with dupe code in the native. John did a PR to unify them in `Eval` but an external client doesn't care about gas, and is already in `PactValue`, so I manually merged it into this with the `Term n -> PactValue` case, and `Eval` factored out.

Reply:
As discussed on slack, changing to accept `Guard PactValue` with native code doing coerce to drop the PactValue coerce fun.